rfcs/0008-custom-event-metadata-from-annotations/README.md

```diff
@@ -0,0 +1,199 @@
+
+For example:
+
+```yaml
+apiVersion: helm.toolkit.fluxcd.io/v2
+kind: HelmRelease
+metadata:
+  name: podinfo
+  namespace: flux-system
+  annotations:
+    event.toolkit.fluxcd.io/image: ghcr.io/stefanprodan/podinfo:latest # {"$imagepolicy": "flux-system:podinfo"}
+spec:
+  chart:
+    spec:
+      chart: podinfo
+      sourceRef:
+        kind: HelmRepository
+        name: podinfo
+  values:
+    image:
+      tag: latest  # {"$imagepolicy": "flux-system:podinfo:tag"}
+```
+
+In this example image-automation-controller will update the image and tag near the markers. If, for example, it
+updates the image to `ghcr.io/stefanprodan/podinfo:6.5.0`, then it will cause notification-controller to start
+propagating events like the one below (most fields omitted for brevity):
+
+```json
+{
+  "involvedObject": {
+    "apiVersion": "helm.toolkit.fluxcd.io/v2",
+    "kind": "HelmRelease",
+    "name": "podinfo",
+    "namespace": "flux-system",
+    "uid": "7d0cdc51-ddcf-4743-b223-83ca5c699632"
+  },
+  "metadata": {
+    "image": "ghcr.io/stefanprodan/podinfo:6.5.0"
+  }
+}
+```
+
+### Alternatives
+
+#### Alternative 1
+
+An alternative for specifying custom metadata fields in Flux objects for sending on events
+is defining `.spec` APIs for such, like `.spec.eventMetadata` available in the Alert API.
+This alternative is not great because:
+
+* Such APIs would be fairly redundant with the well-known Kubernetes annotations.
+* Technically speaking, it is much easier to implement an alternative where the
+field storing the custom metadata is the same and is already available across all the
+Flux objects rather than introducing a new API.
+
+In the specific case of the Alert API this field was introduced because the Alert API
+is obviously a special one in the context of events and alerting. In particular, the
+Alert objects do not generate events themselves, but rather serve as an aggregation
+configuration for matching and propagating events from other Flux objects.
+
+#### Alternative 2
+
+Instead of introducing a new API Group, i.e. `event.toolkit.fluxcd.io`, we could use the API
+Group `notification.toolkit.fluxcd.io` for the same purpose. This alternative is not great
+because it emphasizes an exclusive relationship with notification-controller, which is not the
+case. The events here are also Kubernetes Events, and an API Group that is more general and
+closer to Kubernetes Events is more appropriate.
+
+## Design Details
+
+All Flux controllers use our implementation of the `EventRecorder` interface from the Go
+package `k8s.io/client-go/tools/record`. The implementation hijacks the `AnnotatedEventf()`
+logic to send events not only to Kubernetes but also to notification-controller.
+
+To support the use cases discussed here we would modify the implementation to look
+for annotations prefixed with `event.toolkit.fluxcd.io/` in the Flux objects and
+send them alonside the other metadata of the event.
```

Comment:
The `AnnotatedEventf(object runtime.Object, annotations map[string]string, eventtype, reason string, messageFmt string, args ...interface{})` function signature would stay unchanged.

Inside the function we'll extract only the key/values with the `event.toolkit.fluxcd.io/` prefix from `object.Annotations` and we'll append them as they are to the `annotations` map. NC will receive in the `annotations` map both the controller specific values and the object specific ones. NC will filter and remove the prefix as it does today, the only change to NC would be adding the  `event.toolkit.fluxcd.io/` prefix to its allow list.

Reply:
Yep, what Stefan said above. However, we should probably decide something about the precedence. My proposal:

Do the same we already do for `spec.eventMetadata` in the Alert API: https://fluxcd.io/flux/components/notification/alerts/#event-metadata

> If a field would override one already present on the original event as generated by the emitter, then the override doesn’t happen, i.e. the original value is preserved, and an info log is printed.

Let's keep all metadata that was added by the originating controller (+ spec.summary and spec.eventMetadata on the Alert object) and print an info log on collision. This option prioritizes backwards compatibility. What do you guys think?

Reply:
So:

1. controller metadata
2. Flux object annotations with event.toolkit.fluxcd.io/
3. spec.eventMetadata from Alert object

Where will spec.summary from Alert API go?

Reply:
> Please also mention about introducing this new constant for the prefix and where, as I incorrectly assumed pkg/apis/event, again because I lost the familiarity around this topic.

We have a dedicated file for event meta here: https://github.com/fluxcd/pkg/blob/main/apis/event/v1beta1/metadata.go We should use this instead of apis/metadata for sure!

Reply:
I propose the following override logic, top bottom, last wins:

- User-defined metadata on the alert (set with `.spec.eventMetadata`)
- User-defined summary on the alert (set with `.spec.summary`)
- User-defined metadata on objects (set with `event.toolkit.fluxcd.io` in the event payload)
- Controller metadata (set with `<controller group>.toolkit.fluxcd.io` in the event payload)

The user-defined metadata set on objects will override the one set in the Alert. The controller injected metadata is never altered by an user, as these values are critical to downstream e..g. the commit SHA, digest, chart version, etc.

Reply:
I agree with the priority above, but today `.spec.summary` has absolute priority over everything else:

https://github.com/fluxcd/notification-controller/blob/8b1d9a1e0b5a889243c463a0d1d5f9c980f5ab28/internal/server/event_handlers.go#L421-L446

Reply:
fixed everything, thanks! really good comments and discussion :+1:

Reply:
I forgot to mention where the new API Group would be introduced, I added that now :+1:

Reply:
That makes sense to me. It feels like alert-level metadata should have more priority than object-level, because alert-level will often mean cluster-level in practice. Counter proposal of precedence order from weaker to stronger:

- User-defined metadata on objects (set with `event.toolkit.fluxcd.io` in the event payload)
- User-defined metadata on the alert (set with `.spec.eventMetadata`)
- User-defined summary on the alert (set with `.spec.summary`)
- Controller metadata (set with `<controller group>.toolkit.fluxcd.io` in the event payload)

Reply:
@stefanprodan thoughts?

Reply:
I see some misconceptions, first `summary` used to be the only field for adding custom info, that's no longer the case, we added `eventMetadata` for more structure e.g. having dedicate fields for cluster region/id/etc. 2nd the cluster admins play no role in alerting, Flux multi-tenancy means that Alerts are a concern of the tenant. I see `eventMetadata`  as way to set defaults and object metadata as way to override/refine them.

Reply:
> I see some misconceptions, first summary used to be the only field for adding custom info, that's no longer the case, we added eventMetadata for more structure e.g. having dedicate fields for cluster region/id/etc.

:+1: No misconception here, that's exactly what I said in my message above: `summary` came first, `eventMetadata` came second as an improvement.

> 2nd the cluster admins play no role in alerting, Flux multi-tenancy means that Alerts are a concern of the tenant. I see eventMetadata/summary as way to set defaults and object metadata as a way to override/refine them.

For this argument to have strength I believe we need to think about specific metadata keys that people would want to override. As Sunny mentioned above, `.spec.summary` has been the solution to identify clusters and I'm sure people wouldn't want to override that. Is there a specific key that you picture users wanting to override? I'm fine with this choice if we can come up with a few examples.

Reply:
Also I'm realizing that we definitely need to document all of that in the RFC as for the chosen precedence order not to seem arbitrary for the reader, we need specific examples there to justify our decision.

Reply:
Summary is a good example of something that you would one to override, for example the Alert would have a summary like "Some app is failing (link to incident response playbooks)" and the object would have "This exact app is failing and here how you fix it (link to the app incident response playbook)".

Reply:
Agreed, then `summary` is a key that users may want to override. Would users also want to override cluster info from `.spec.eventMetadata`? So far it looks like the ideal precedence order is:

- User-defined summary on the alert (set with `.spec.summary`)
- User-defined metadata on objects (set with `event.toolkit.fluxcd.io` in the event payload)
- User-defined metadata on the alert (set with `.spec.eventMetadata`)
- Controller metadata (set with `<controller group>.toolkit.fluxcd.io` in the event payload)

Reply:
By the way, we need to update the docs of `.spec.summary`: https://fluxcd.io/flux/components/notification/alerts/#summary

> `.spec.summary` is an optional field to specify a short description of the impact and **affected cluster**.

It is conflicting with https://fluxcd.io/flux/components/notification/alerts/#event-metadata, which is the one ideal for cluster info.

We need to remove the "and affected cluster" part.

Reply:
We could deprecate/remove summary in Alert v1 GA, now that we have `eventMetadata`, the summary is redundant IMO.

Reply:
Agreed 100% with that, `.spec.summary` is garbage now!

Reply:
Anyway deprecating summary is outside the scope of this RFC. So we should include it in the override rules.

Reply:
Yep, I'm doing that right now :ok_hand:

Reply:
fixed, thanks again! @darkowlzz I'll let you press the `Resolve conversation` button whenever you're ready :ok_hand:

Reply:
> Because Stefan expressed that metadata from the objects should take the highest precedence, it seemed like the obvious next conclusion would be to deprecate both of the alert summary and event metadata, as the objects can add their custom metadata natively now.

No, this is not the next conclusion at all, no idea how you got this impression. What I was trying to say is that `.spec.summary` can be deprecated in favour of `.spec.eventMetadata.summary`. I see a lot of value in setting common metadata in the Alert e.g. cluster region/name/etc and app specific metadata in the object e.g. version/image/tag/etc with the option for the app specific values to override the generic ones e.g. `.metadata.annotations.summary` overrides `eventMetadata.summary`.

Reply:
I share the same opinion as above, but if the object-level metadata is supposed to override anything from `.spec.eventMetadata` we need to be very specific about it, our last agreed precedence order does not specify that:

- User-defined summary on the alert (set with `.spec.summary`)
- User-defined metadata on objects (set with `event.toolkit.fluxcd.io` in the event payload)
- User-defined metadata on the alert (set with `.spec.eventMetadata`)
- Controller metadata (set with `<controller group>.toolkit.fluxcd.io` in the event payload)

If object annotations can override `.spec.eventMetadata.summary` then the updated precedence order would be:

- User-defined summary on the alert (set with `.spec.summary` or `.spec.eventMetadata.summary`)
- User-defined metadata on objects (set with `event.toolkit.fluxcd.io` in the event payload)
- User-defined metadata on the alert (set with `.spec.eventMetadata`)
- Controller metadata (set with `<controller group>.toolkit.fluxcd.io` in the event payload)

Being specific about exceptions like this is important because we agreed that `.spec.eventMetadata` is mainly targeted towards cluster-level metadata and hence would have priority over object annotations. I created a section Precedence Order in the RFC which is very specific about everything, please take a look.

Reply:
Thanks Sunny, I agree with this reasoning. If we are going to deprecate `.spec.summary` for Alert API v1 GA then it doesn't make sense to design the feature with it in the center. If I understand your last message correctly I believe we are back to the following precedence order until we remove `.spec.summary`:

- User-defined metadata on objects (set with `event.toolkit.fluxcd.io` in the event payload)
- User-defined metadata on the alert (set with `.spec.eventMetadata`)
- User-defined summary on the alert (set with `.spec.summary`)
- Controller metadata (set with `<controller group>.toolkit.fluxcd.io` in the event payload)

This looks good to me, I have already posted a long message with a reasoning that makes sense to me for backing this choice.

@stefanprodan seemed to have a strong opinion against this order, but I believe that with our recent unanimous agreement on deprecating `.spec.summary` it shouldn't be a problem, the final precedence order after the removal of `.spec.summary` is the same Stefan and I agreed upon earlier. In any case, I'll let Stefan raise any other concerns or resolve the thread.

I have a final question about the deprecation of `.spec.summary`: does deprecation here mean removing support for it, or just adding a deprecation warning in the struct field comment and in the docs? I believe we could do the latter in Flux 2.5 and remove it in Flux 2.x where x is the version of Alert API v1 GA.

Reply:
> Other than this part, we still need to add some more details about the changes in the event server to accommodate the new behavior. Maybe a few things about moving things around to allow for metadata ordering based on the metadata key prefixes, just conceptually in terms of middlewares and event/request handler, and any other thing that you can think of.

The implementation in notification-controller is not part of defining any public APIs, neither is related to defining software behavior that must be documented for users. We should keep this discussion for the n-c PR.

Reply:
All the rest is already addressed, please take a look.

Reply:
> The design details were short and vague before. Had I not pointed out the shallow details, we wouldn't have considered all the details that have been added now.

I'm super grateful for your insistence, it has truly improved the quality of the RFC. I can't thank you enough. Thank you, Sunny.

Can we please just stop posting in this thread? Future readers will have a horrible time trying to read or find anything inside it. A thread is for a single subject, but in this one we talked about countless subjects.

I'm just asking to continue this specific discussion around fine-grained implementation details in notification-controller in a more appropriate place, there's no unwillingness to discuss any concerns here. If you have anything else that makes sense to be discussed here, I'm more than willing to do so. I just think that any outcomes from this specific discussion would/should not change the content of the RFC, and hence the RFC is not the right place to have that conversation. Please agree with that. Please consider how flexible we were addressing all your concerns so far. Let's just discuss small notification-controller details [here](https://github.com/fluxcd/notification-controller/pull/1014), please.